libraries/AP_Baro/AP_Baro_atmosphere.cpp

```diff
@@ -0,0 +1,388 @@
+#include "AP_Baro.h"
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  atmospheric model options
+ */
+
+#ifndef HAL_BARO_ATMOSPHERIC_TABLE
+// default to using tables when doing double precision EKF (which implies more flash space and faster MCU)
+// this allows for non-table testing with --ekf-single configure option
+#define HAL_BARO_ATMOSPHERIC_TABLE HAL_WITH_EKF_DOUBLE
+#endif
+
+/*
+  return altitude difference in meters between current pressure and a
+  given base_pressure in Pascal. This is a simple atmospheric model
+  good to about 11km AMSL. Enable the table based function for higher altitudes
+*/
+float AP_Baro::get_altitude_difference_function(float base_pressure, float pressure) const
+{
+    float ret;
+    float temp    = get_ground_temperature() + C_TO_KELVIN;
+    float scaling = pressure / base_pressure;
+
+    // This is an exact calculation that is within +-2.5m of the standard
+    // atmosphere tables in the troposphere (up to 11,000 m amsl).
+    ret = 153.8462f * temp * (1.0f - expf(0.190259f * logf(scaling)));
+
+    return ret;
+}
+
+#if HAL_BARO_ATMOSPHERIC_TABLE || CONFIG_HAL_BOARD == HAL_BOARD_SITL
+/*
+  MIL-STD 3013 (1962 U.S. Standard Atmosphere) in Tabular Form.
+
+  Using this table is more accurate over a much wider range of
+  altitudes than using the function
+ */
+static const struct {
+    float amsl_feet;
+    float pressure_Pa;
+    float density;
+    float temp_K;
+} atmospheric_table[] = {
+    {-15000, 169732.9, 1.86000, 357.14 },
```

Comment:
This graphs that @davidbuzz seem to show why high-order polynomials didn't work well - discontinuities!  - but splitting based on domain would seem to work if a function were to be created.

Reply:
the temperature in my PR was totally wrong due to an error extracting the table. This didn't impact on the flight tests as temperature isn't needed